std/array.d

```diff
@@ -796,22 +796,49 @@ private void copyBackwards(T)(T[] src, T[] dest)
 
     assert(src.length == dest.length);
 
-    if (!__ctfe || hasElaborateCopyConstructor!T)
+    static if (!hasElaborateAssign!T)
     {
-        /* insertInPlace relies on dest being uninitialized, so no postblits allowed,
-         * as this is a MOVE that overwrites the destination, not a COPY.
-         * BUG: insertInPlace will not work with ctfe and postblits
-         */
-        memmove(dest.ptr, src.ptr, src.length * T.sizeof);
+        if (!__ctfe)
+        {
+            /* insertInPlace relies on dest being uninitialized, so no postblits allowed,
+             * as this is a MOVE that overwrites the destination, not a COPY.
+             * BUG: insertInPlace will not work with ctfe and postblits
+             */
+            memmove(dest.ptr, src.ptr, src.length * T.sizeof);
+            return;
+        }
     }
-    else
+
+    immutable len = src.length;
+    for (size_t i = len; i-- > 0;)
     {
-        immutable len = src.length;
-        for (size_t i = len; i-- > 0;)
-        {
-            dest[i] = src[i];
+        dest[i] = src[i];
+    }
+}
+
+unittest
+{
+    static struct S
+    {
+        uint a;
+        this(this) {
+            a++;
         }
     }
+
+    static auto foo(S[] a)
+    {
+        auto b = new S[a.length];
+        copyBackwards(a, b);
+        return b;
+    }
+
+    enum a = new S[8];
+
+    auto r = foo(a);
+    assert(r[0].a == 1);
+    enum c = foo(a);
+    static assert(c[0].a == 1);
```

Comment:
Fair enough,  but 

>  Classes and unions never have elaborate assignments.

`insertInPlace` isn't tested for these,  maybe worth adding?